### PR TITLE
[ROCm][DSv4] Fix sparse-indexer logits collapse on gfx950/gfx942

### PR DESCRIPTION
**Pull request by frida-andersson**
On gfx950 and gfx942, AITER's `deepgemm_fp8_paged_mqa_logits` mis-scores the DSv4 sparse indexer during decode (wrong paged-cache layout), collapsing long-context retrieval. `VLLM_DSV4_LOGITS_FIX=1` (off by default) reroutes decode + MTP through a Triton kernel instead

The replacement kernel (`_fp8_paged_mqa_logits_decode_kernel`) reads the fp8 KV cache in its true block-flat layout and computes `relu(q·k)·weight·scale` per key position, splitting each query's key range across programs so there is no cross-program reduction or host sync — which keeps it capturable in a full CUDA graph

The decode kernel computes its q·k dot directly in fp8 (fp8 MFMA with f32 accumulation) rather than upcasting q and k to f32 first. Against that f32-upcast path the dot is 4.4–4.7× faster on gfx942 and up to ~9× on gfx950 at serving batch sizes, for the same result — bit-identical on gfx942, and on gfx950 within a ~1e-5 rounding difference that leaves the selected keys unchanged (GSM8K, RULER and MTP all agree).

Env: vLLM `0.26.1rc1.dev124`, AITER `0.1.16.post5`, DeepSeek-V4-Flash, TP=4, fp8 KV, `block_size=256`, `max_model_len=40960`, 4x gfx950

## RULER NIAH

Accuracy % as `stock → fix` (50 samples/cell, temp=0, concurrency 16)

| context | niah_1 | niah_2 | niah_3 |
|--:|:--:|:--:|:--:|
| 4K  | 62 → 90 | 60 → 58 | 8 → 54 |
| 8K  | 40 → 96 | 18 → 66 | 10 → 38 |
| 16K | 24 → 86 | 16 → 68 | 4 → 40 |
| 32K | 12 → 92 | 0 → 80  | 2 → 48 |

At 4K the indexer prunes little, so the fix has little to correct there; niah_2 `60 → 58` is within run-to-run noise. The fix's effect grows with context, where stock collapses and the fix holds

As a sanity check, I also ran niah_1 through the torch reference - the slow but known-correct version of the decode-logits calculation that the fix's Triton kernel is designed to reproduce. It recovered accuracy to 81% at 16K and 92% at 32K, essentially matching the fix's own niah_1 numbers

## GSM8K

flexible-extract accuracy (1319 problems), prefix caching disabled. The gap widens with shot count: more few-shot context means the indexer prunes more, so stock's mis-ranking costs more and the fix recovers more. 5-shot is too short to engage the indexer (within noise)

| shots | stock | fix | Δ |
|--:|--:|--:|--:|
| 5  | 0.9477 | 0.9492 | +0.002 |
| 20 | 0.9212 | 0.9553 | +0.034 |
| 60 | 0.8772 | 0.9507 | +0.074 |

## Throughput

Decode throughput on the committed kernel (with the fp8 dot), prefix caching disabled, 512 output tokens, `ignore_eos`, 1 warmup + 2 reps. `stock` = AITER fused paged-MQA-logits kernel, `fix` = Triton replacement

| in / conc | stock tok/s | fix tok/s | stock TPOT | fix TPOT |
|--|--:|--:|--:|--:|
| 16K / 1  | 72.5  | 74.8  | 12.72 | 12.29 |
| 16K / 32 | 652.8 | 696.9 | 41.26 | 37.93 |
| 32K / 1  | 69.3  | 69.0  | 12.28 | 12.35 |

The fix runs at parity with AITER's fused kernel — differences are within run-to-run noise, since the logits step is only a few percent of decode

### Kernel occupancy tuning

The kernel is memory-bound over the KV range, so the number of KV-range splits per query sets occupancy. A fixed 32 leaves most CUs idle at low batch / long context, so it scales as `min(256, tiles, 1024 / rows)` (all static at launch, so still CUDA-graph-safe). Isolated kernel latency vs the fixed-32 version, output unchanged:

| batch | ctx | fixed 32 | tuned | speedup |
|--:|--:|--:|--:|--:|
| 1  | 16K | 62.9 us | 34.9 us | 1.8x |
| 1  | 32K | 115  us | 36.5 us | 3.2x |
| 8  | 16K | 64.4 us | 60.5 us | 1.06x |
| 32 | 16K | 200  us | 192  us | 1.04x |

End-to-end the effect is small — ~3-5% TPOT at concurrency 1, within noise at c8/c32 (where the split count is back to 32). No regression anywhere

## gfx942 (MI300-class): same bug, same fix

The identical collapse reproduces on gfx942 and the Triton reroute fixes it there too. The kernel is unchanged — the fp8 KV-cache layout and torch reference are identical across both archs — so only the arch gate widens (`_ON_GFX950` → `_ON_GFX950 or _ON_GFX942`). RULER NIAH `stock → fix` (TP8, fp8 KV, 8x gfx942, 50 samples/cell, temp=0, concurrency 16):

| context | niah_1 | niah_2 | niah_3 |
|--:|:--:|:--:|:--:|
| 4K  | 60 → 86 | 64 → 84 | 14 → 56 |
| 8K  | 44 → 92 | 26 → 60 | 6 → 54  |
| 16K | 22 → 76 | 14 → 68 | 4 → 52  |
| 32K | 6 → 72  | 4 → 82  | 0 → 58  |

Same signature as gfx950: stock collapses toward ~0 by 32K, the fix recovers and holds. MTP (`num_speculative_tokens=3`) verified, consistent with the non-MTP fix

### CODE REVIEW

**Comment by hellozhuo-amd** [comment]
I have reviewed the code under [ticket](https://amd.atlassian.net/browse/SILOTIGER-868). 
The original fp32 tl.dot was changed to fp8 version, and additional assert regarding   head_size was added. Performance and accuracy looks sound, by comparing with the stock kernel and the torch reference respectively.
The code looks good to me!
A few things to mention:
1.  "VLLM_DSV4_LOGITS_FIX" causes additional env var to the original vllm, we may need additional changes in the code to avoid that (but let's first wait for the outputs from the vllm experts).
2.  There is a gap in accuracy (86% at 16K context length and 92% at 32K from the PR) compared with SGLang (100% at 16K and 98% at 32K). However, the accuracy matches the numbers using torch reference. This may need to be further investigated (i.e., vLLM vs. SGLang).

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 1]
```diff
@@ -25,6 +25,14 @@
     _ON_GFX950 = False
 
 
+import os as _os
+
+# gfx942/gfx950: AITER's paged MQA-logits decode kernel scores the sparse
+# indexer incorrectly and collapses long-context retrieval. Opt in to the Triton
+# path to fix.
+_DSV4_LOGITS_FIX = _os.environ.get("VLLM_DSV4_LOGITS_FIX", "0") == "1"
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Let's not add a flag here. If this version requires a fix like this, we use it immediately. We want to make sure OOTB experience without the need to specify too many flags.

**Comment by frida-andersson** [comment]
Thanks @tjtanaa for the review! I've addressed the comment now by removing the `VLLM_DSV4_LOGITS_FIX` env var so the Triton indexer path is now always-on for gfx942/gfx950 (no flags needed). Re-validated that it matches the torch reference and serves cleanly on the standard recipe. PTAL when you get a chance

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -504,6 +655,19 @@ def rocm_fp8_paged_mqa_logits(
     """
     from vllm._aiter_ops import rocm_aiter_ops
 
+    # gfx942/gfx950: AITER's paged MQA-logits decode kernel scores the sparse
+    # indexer incorrectly and collapses long-context retrieval, so route the
+    # decode/MTP path through the Triton kernel instead.
+    if _ON_GFX950 or _ON_GFX942:
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: So AITER's paged MQA-logits decode kernel has issue for all cases? If we do this, we will be introducingperf regression to all models as it is going to use triton or even torch, rather than gluon kernel from aiter.

Comment (frida-andersson) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Thanks for the comment! AITER's kernel is fine for the SHUFFLE layout DSv32 uses, the bug is DSv4-specific, where the compressor writes block-flat but `deepgemm_fp8_paged_mqa_logits` reads shuffled. This only hits sparse-indexer decode, not other models. The gating is too wide though, I'll tighten it so DSv32/GLM keep the current gluon path

**Comment by frida-andersson** [comment]
The original gating was too broad, I've modified it in [2deb753](https://github.com/vllm-project/vllm/pull/50455/commits/2deb7531ee37e5d721d365203b0d69f440a58d54). Triton decode logits now only when `skip_k_cache_insert` (DSv4 block-flat compressor path), DSv32/GLM stay on AITER gluon. Trace + GSM8K 95.2% on DSv4-Pro TP8. PTAL @tjtanaa

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 3]
```diff
@@ -504,11 +657,24 @@ def rocm_fp8_paged_mqa_logits(
...
+            return rocm_fp8_paged_mqa_logits_triton(
+                q_fp8, kv_cache_fp8, weights, context_lens, block_tables, max_model_len
+            )
+        # Non 64-aligned page size (not used in prod): eager torch ref.
+        return fp8_paged_mqa_logits_torch(
+            q_fp8, kv_cache_fp8, weights, context_lens, block_tables, max_model_len
+        )
+
+    aiter_paged_mqa_logits_module = None
+    # if rocm_aiter_ops.is_enabled():
```
Comment (tjtanaa) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Let's remove this line of comment

Comment (frida-andersson) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Removed the stale comment in [7529cc0](https://github.com/vllm-project/vllm/pull/50455/changes/100bda95a02980f608ef67b2f85df33aa585b8c7). PTAL @tjtanaa

**Review by tjtanaa — APPROVED** [review]
LGTM.

**Comment by frida-andersson** [comment]
@tjtanaa since your LGTM I folded the CodeRabbit follow-ups plus an optimization that skips the -inf pad into 6304894 (decode top-k already stops at seq_lens). Prefill gather is NORMAL, weights use stride. Flash NIAH now at 100% through 32K, tok/s flat/slightly up. PTAL

**Comment by frida-andersson** [comment]
AMD CI red is already failing on main, unrelated to this change

**Comment by frida-andersson** [comment]
Noticed some issues with DSv32 - converting to draft while working on it

**Comment by peizhang56** [comment]
Hi @frida-andersson 
Any issues with DSv4-Flash-0731?

**Comment by peizhang56** [comment]
This PR fixed an accuracy issue for one of our customers with DeepSeek V4 Flash 0731.
They are interested in when will this be merged.

**Comment by frida-andersson** [comment]
@tjtanaa the `skip_k_cache_insert` gate was still too broad as DSv3.2’s ROCm writer sets that too, so it was hitting the Triton kernel even though its cache is shuffled, not block-flat. I switched it to `compress_ratio > 1` in eb09a3c. That arg was already on the indexer op: DSv4’s C4A indexer passes 4, DSv3.2 never sets it (stays 1). GLM goes through kpool and also doesn’t pass it, so it stays on AITER. 

Flash NIAH is 100% through 32K, DSv32 GSM8K stays at stock (~0.96). PTAL

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 4]
```diff
@@ -598,6 +602,157 @@ def fp8_paged_mqa_logits_torch(
...
+    (out_logits,) = current_workspace_manager().get_simultaneous(
+        ((batch_size * next_n, max_model_len), torch.float32),
+    )
+
+    # Memory-bound over the KV range: split each row's keys across programs so
+    # few-row / long-context launches still fill the GPU. All terms are static
+    # at launch, so the grid stays CUDA-graph-safe.
+    rows = batch_size * next_n
+    tiles_cap = (max_model_len + BLOCK_KV - 1) // BLOCK_KV
+    N_SPLITS = max(1, min(256, tiles_cap, 1024 // rows))
```
Comment (simondanielsson) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: **Question**: I might be off here, but would we benefit on gfx942 from using _decode_cu_count here instead of 256?

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 5]
```diff
@@ -598,6 +602,157 @@ def fp8_paged_mqa_logits_torch(
...
+        # (accum order), top-k unchanged.
+        s = tl.dot(kv, tl.trans(q))  # [BLOCK_KV, H]
+        s = tl.maximum(s, 0.0)
+        s = s * w[None, :]
+        s = tl.sum(s, axis=1)  # [BLOCK_KV]
+        s = s * sc
+        tl.store(logits_ptr + row * stride_logits_row + pos, s, mask=mask_pos)
+
+
+def rocm_fp8_paged_mqa_logits_triton(
```
Comment (simondanielsson) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:687]: **Suggestion**: Is it worth adding a UT for this to ensure correctness?

**Review by simondanielsson — COMMENTED** [review]
Nice work :)

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 4 above)
Comment (frida-andersson) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py]: Good catch! I initially created the PR for gfx950 but it makes sense to switch to `_decode_cu_count()` since it's gfx942 too

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py (hunk 5 above)
Comment (frida-andersson) [reply vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:687]: Added `test_fp8_paged_mqa_logits_triton_matches_torch_ref`, Triton vs the torch ref on the valid [:seq_len] region. Passes on gfx950

**Comment by frida-andersson** [comment]
@tjtanaa ping for re-review. AMD CI is red on the NIXL test _failed_recv_reqs, same failure is on today’s main nightly ([build 12959](https://buildkite.com/vllm/amd-ci/builds/12959))

**Comment by frida-andersson** [comment]
@simondanielsson @tjtanaa - gate change since your last look: Triton/NORMAL only when compress_ratio == 4 (V4.0 C4A). V4.1-Flash is SHUFFLE at ratio 1 and 2; the old >1 check sent ratio-2 layers to the wrong reader. Verified gfx950 TP4: V4-Flash → Triton C4A, V4.1 and DSv3.2 → AITER. PTAL

**Review by tjtanaa — APPROVED** [review]
LGTM
